src/guildenstern/guildenserver.nim

```diff
@@ -111,20 +111,14 @@ proc handleRead*(gs: ptr GuildenServer, data: ptr SocketData) =
 proc closeSocket*(ctx: Ctx) {.raises: [].} =
   try:
     when defined(fulldebug): echo "closing ctx socket: ", ctx.socketdata.socket
```

Comment:
This is really the crux of the issue that I was witnessing before.  Sockets would get randomly closed, leading to unexpected "connection reset" type of errors on the client.  

As the code now mentions, `ctx.socketdata` gets cleared during the call to `Selector.unregister()`, which led to closing file descriptor 0 (which is stdin initially) but later FD 0 gets reallocated with other socket connections, and those connections would get randomly closed when a socket would time out, or when an invalid request was sent to the server.